### DISCUSSION TITLE
`signal` in atom unexpected behavior

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/jotai#2028] [General] [answered]
## Summary

```tsx
const idAtom = atom(1);

const countAtom = atom(async (get, { signal }) => {
  const id = get(idAtom);
  console.log("connect", id);
  signal.addEventListener("abort", () => {
    console.log("disconnect", id); // <-- never called
  });
  return new Promise((resolve) => setTimeout(() => resolve(id), 100));
});
```

## Link to reproduction

https://codesandbox.io/s/trusting-phoebe-x5q6v7?file=/src/App.tsx

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by himself65** [comment]
Our production code wants to move provider connection into `atom` instead of useEffect.

But the weird thing is abort signal emits `abort` immediately after it returns the value, which makes our connection lost.

Refs: https://github.com/toeverything/AFFiNE/pull/3075

**Comment by himself65** [comment]
> Our production code wants to move provider connection into `atom` instead of useEffect.
> 
> But the weird thing is abort signal emits `abort` immediately after it returns the value, which makes our connection lost.
> 
> Refs: [toeverything/AFFiNE#3075](https://github.com/toeverything/AFFiNE/pull/3075)

I think `connect` should emit twice and `disconnect` once. But it now only 1 1

![image](https://github.com/pmndrs/jotai/assets/14026360/0a5309a4-d7fe-40f4-95c5-ebe3a53f1fa4)
![image](https://github.com/pmndrs/jotai/assets/14026360/7145100a-e51f-4abe-8721-ab40cee681cf)

**Comment by dai-shi** [comment, marked as answer]
For now, we only `abort()` if the returned promise is still "pending".

We generally don't recommend creating a new connection in the read function. It must be pure and expects everything to be garbage collectable.

**Reply by himself65** [reply]
Thanks for the suggestion. But I still think `signal` should call abort even if the promise has resolved because it allows the atom to clean up some `side effect.` I think the async atom cannot be pure 100%. The async atom is meant to use `Suspense` in React, so it must have the side effect(or we say algebraic effect). I know what you're talking about, but I still hope it could support my purpose

My workaround is to use `rxjs` now.

```
const myAtom = atomWithObservable(get => {
  return new Observable(sub => {
    const controller = new AbortController();
    const signal = controller.signal;
    // do sth async
    sub.next(value)

    return () => controller.abort()
  })
})
```

**Reply by dai-shi** [reply]
Related discussion: #2025

It's a bigger issue. AbortSignal wouldn't solve all cases anyway (besides its purpose is different.)

"pure" is a bit difficult term. Here's a reference: https://gist.github.com/sebmarkbage/75f0838967cd003cd7f9ab938eb1958f

I think "garbage collectable" would be appropriate in this context.

**Reply by wootra** [reply]
I couldn't understand why we need abort signal when it is not even pending since we can use then() or catch() function to clean up side effects. 
but soon realized that the Promise is resolved (or catch) even though it is aborted.
Based on that, I experimented if this code is working, and looks like it is working(https://codesandbox.io/s/reverent-ioana-sr3kxk?file=/src/App.tsx)

```
const countAtom = atom(async (get, { signal }) => {
  const id: number = get(idAtom);

  signal.onabort = () => {
    console.log("abort triggered", id);
  };

  return new Promise((resolve, reject) => {
    setTimeout(() => (id % 3 === 0 ? reject(id) : resolve(id)), 1000);
  })
    .then(() =>
      signal.aborted // <=== we can check signal.abort in the .then or .catch function
        ? console.log(`${id} Promise is aborted(resolve)`)
        : console.log(`${id} Promise is resolved`)
    )
    .catch((e) =>
      signal.aborted
        ? console.log(`${id} Promise is aborted(reject)`)
        : console.error("error", e)
    );
});
```
So, we can safely clean up only when is aborted.
you can confirm the behavior when you click the button twice fast (before 1sec) and check the console

**Reply by himself65** [reply]
Yeah, your code is make lot of sense

**Reply by dai-shi** [reply]
Nice. Feel free to improve docs. You can also do `clearTimeout` with `.onabort`.
